bazel/protobuf.patch

```diff
@@ -21,3 +21,17 @@ index efc3d8e7f..746ad4851 100644
 
  ################################################################################
  # Protobuf Runtime Library
+diff --git a/python/google/protobuf/__init__.py b/python/google/protobuf/__init__.py
+index 97ac28028..8b7585d9d 100644
+--- a/python/google/protobuf/__init__.py
++++ b/python/google/protobuf/__init__.py
+@@ -31,3 +31,9 @@
+ # Copyright 2007 Google Inc. All Rights Reserved.
+
+ __version__ = '3.14.0'
```

Comment:
Do we need to maintain a patch or can we get a change landed upstream to avoid this? CC @envoyproxy/dependency-shepherds

Reply:
Honestly, I don't quite understand what's the deal with google dependencies, python, and bazel. If someone who understands it better could pick it up, I'd appreciate it. I only see it at the level that this was there before, and it's gone, and it works with it.

Reply:
I think this is from https://github.com/protocolbuffers/protobuf/commit/66e3562aafce093d30473555248f22b3fdc88aad. @dlj-NaN do you know why that PR would have caused `google.api` imports to start breaking?

Reply:
One thing to keep in mind is that I can't update googleapis because some python rules are being removed (I need to find out why). So there's that.

Reply:
OK. I don't think any of our build experts has a much better grasp of the Python/Bazel/GoogleAPIs internals, I'd probably start from first principles debugging :(

Reply:
All I wanted was to update cel-cpp which needs new protobuf. I ended up fixing python, golang, and windows builds... From the bug description it just sounds that how python protobufs are packaged is fundamentally broken since they cross python module boundaries. I don't what's the right thing to do here.

Reply:
protocolbuffers/protobuf@66e3562 gets pretty deep into the semantics of Python imports and different packaging. The interaction addressed by that commit is very tricky, and may be hidden by varying the order of imports.

(I'm guessing the reason this appeared with a protobuf commit was exactly due to order of imports... `google.auth` comes after `google.api` alphabetically, but due to transitive imports, `google.protobuf` is effectively a very early import in many scenarios.)

The upshot is that there shouldn't be any files named `google/__init__.py` in other installed packages. Just as a first guess, this might be caused by installing googleapis/google-auth-library-python as an egg (maybe via `setup.py install`?), but other `google` packages using native namespaces. (This also gets into what Bazel does in its pip rules....)

Anyhow, some fixes might be:

1.  Remove `google/__init__.py` from googleapis/google-auth-library-python. (@busunkim96 to comment on that.)
2.  Pre-install the package with an up-to-date version of `pip`, which _should_ install using the native namespace.
3.  Update Bazel's pip rules to make sure it does (2).

Reply:
> Remove google/__init__.py from googleapis/google-auth-library-python. 

Removing `google/__init__.py` from the `google-auth` will probably be done next quarter as we officially drop 2.7.  google-auth is last on the list since it is transitively required by nearly all the cloud client libraries.

Over the past year we've gradually moved `google-cloud-API` packages to using native namespace packages (example: [google-cloud-texttospeech](https://github.com/googleapis/python-texttospeech/tree/master/google/cloud)) while other libraries like `google-api-core`, `google-cloud-core`, and `google-auth` have kept `google/__init__.py` with pkg_resources style namespace packages. I see now that setuptools [does not recommend](https://packaging.python.org/guides/packaging-namespace-packages/#creating-a-namespace-package) this, but we have not seen customers open issues about it. Our docs always point folks `pip install` packages.

> One thing to keep in mind is that I can't update googleapis because some python rules are being removed (I need to find out why). So there's that.

Are you referring to Bazel rules in https://github.com/googleapis/googleapis or something else?

I don't have much context here so I'm not sure it matters, but the google-cloud-* libraries get their copy of many commonly used protos (`google.api`, `google.type`, ...) from  https://github.com/googleapis/python-api-common-protos, published on PyPI as [google-apis-common-protos](https://pypi.org/project/googleapis-common-protos/)

Reply:
Not sure if this is related but I'm trying to land an update to how we process pip dependencies using the latest `rules_python` at https://github.com/envoyproxy/envoy/pull/14329

Reply:
I opened an issue for further discussion https://github.com/envoyproxy/envoy/issues/14412. 
We're using googleapis/googleapis here because some protos are missing from https://github.com/googleapis/python-api-common-protos (e.g. `google.api.expr` for CEL). Does this mean we can't use PIP approach since there's no pre-generated package I am aware for it?

Reply:
pip handling changes with the latest `rules_python` - https://github.com/bazelbuild/rules_python/releases/tag/0.1.0. PR https://github.com/envoyproxy/envoy/pull/14329 that moves us to that handling is blocked due to Windows (/cc @envoyproxy/windows-dev). In theory pip supports installing from Git - https://pip.pypa.io/en/stable/reference/pip_install/#git but no idea if this works for the current deprecated pip method or the new method we are trying to get to.